### ISSUE TITLE
Labor Tooltip Suggestion

### ISSUE THREAD

**Issue by RusTit** [Dwarf-Therapist/Dwarf-Therapist#94]
A little while ago I made for myself a change in the appearance of the tips.
Do you think it's worth to adapt the code and make a PR?

https://github.com/splintermind/Dwarf-Therapist/compare/DF2016...RusTit:labor_tooltip_feature?expand=1
(Scroll down for screenshots)

**Comment by cvuchener** [comment]
If I understand correctly you were using a two-column table while adding everything in a <pre> element and reinventing an (imperfect) word-wrapping.

Despite the previous changes from #16, dwarf tool-tips can still be too tall and multi-column could be a nice solution. But I don't like your custom word-wrap function. Why is it necessary?

**Comment by RusTit** [comment]
I also do not like my solution, but in fact I did not find a way in Qt to specify the (max) width of the tooltips.
It seemed to me that Qt using platform functions and does not have the ability to customize.

**Comment by cvuchener** [comment]
Table cells supports the width attribute: https://doc.qt.io/qt-5/richtext-html-subset.html#table-cell-attributes I hope the size is in [device independent pixels](https://doc.qt.io/qt-5/highdpi.html#glossary-of-high-dpi-terms), so it can work with high-dpi displays. Or try with css: `style="width: XXXem"`, it is not in the supported list but it would be best to have the width depend on the font size.

**Comment by RusTit** [comment]
I tried the solution you proposed in the first place.
But in the context of the tooltip, it did not work at all (I still think that a simplified code is used for renderer tooltip).

**Comment by RusTit** [comment]
Truth it was a long time, maybe I still made a mistake and went the wrong way.

**Comment by cvuchener** [comment]
It works for me (quick test with MinGW 7.3.0 + Qt 5.10.1):

![dwarf_tooltip](https://user-images.githubusercontent.com/3874173/41114436-b33d424e-6a84-11e8-9d85-5244a2870732.png)

High-dpi scales properly:
![dwarf_tooltip_2x](https://user-images.githubusercontent.com/3874173/41114440-b8e35314-6a84-11e8-9884-882df9553ba0.png)

```diff
diff --git a/src/dwarf.cpp b/src/dwarf.cpp
index a210306d..712996d4 100644
--- a/src/dwarf.cpp
+++ b/src/dwarf.cpp
@@ -2598,7 +2598,7 @@ QString Dwarf::tooltip_text() {
                                     "<ul style=\"margin:0\">%2</ul>"
                                     "</div>");

-    QString title;
+    QString title, col1, col2;
     if(s->value("tooltip_show_icons",true).toBool()){
         title += tr("<b><h3 style=\"margin:0\"><img src='%1'> %2 %3</h3><h4 style=\"margin:0\">%4</h4></b>")
                 .arg(m_icn_gender).arg(m_nice_name).arg(embedPixmap(m_icn_prof))
@@ -2618,37 +2618,37 @@ QString Dwarf::tooltip_text() {
 #endif

     if(s->value("tooltip_show_caste",true).toBool())
-        html.append(item_with_title.arg(tr("Caste:")).arg(caste_name()));
+        col1.append(item_with_title.arg(tr("Caste:")).arg(caste_name()));

     if(m_is_animal || s->value("tooltip_show_age",true).toBool())
-        html.append(item_with_title.arg(tr("Age:")).arg(get_age_formatted()));
+        col1.append(item_with_title.arg(tr("Age:")).arg(get_age_formatted()));

     if(m_is_animal || s->value("tooltip_show_size",true).toBool())
-        html.append(item_with_title.arg(tr("Size:")).arg(QLocale(QLocale::system()).toString(m_body_size * 10) + " cm<sup>3</sup>"));
+        col1.append(item_with_title.arg(tr("Size:")).arg(QLocale(QLocale::system()).toString(m_body_size * 10) + " cm<sup>3</sup>"));

     if(!m_is_animal && s->value("tooltip_show_noble",true).toBool())
-        html.append(item_with_title.arg(tr("Profession:")).arg(profession()));
+        col1.append(item_with_title.arg(tr("Profession:")).arg(profession()));

     if(!m_is_animal && m_pending_squad_id > -1 && s->value("tooltip_show_squad",true).toBool())
-        html.append(item_with_title.arg(tr("Squad:")).arg(m_pending_squad_name));
+        col1.append(item_with_title.arg(tr("Squad:")).arg(m_pending_squad_name));

     if(!m_is_animal && m_noble_position != "" && s->value("tooltip_show_noble",true).toBool())
-        html.append(item_with_title.arg(m_noble_position.indexOf(",") > 0 ? tr("Noble Positions") : tr("Noble Position")).arg(m_noble_position));
+        col1.append(item_with_title.arg(m_noble_position.indexOf(",") > 0 ? tr("Noble Positions") : tr("Noble Position")).arg(m_noble_position));

     if(!m_is_animal && s->value("tooltip_show_happiness",true).toBool()){
-        html.append(item_with_title.arg(tr("Happiness:")).arg(m_happiness_desc));
+        col1.append(item_with_title.arg(tr("Happiness:")).arg(m_happiness_desc));
         if(m_stressed_mood)
             html.append(item_with_title.arg(tr("Mood:")).arg(gdr->get_mood_desc(m_mood_id,true)));
     }

     if(s->value("tooltip_show_orientation",false).toBool())
-        html.append(item_with_title.arg(tr("Gender/Orientation")).arg(m_gender_info.full_desc));
+        col1.append(item_with_title.arg(tr("Gender/Orientation")).arg(m_gender_info.full_desc));

     if(!m_is_animal && !m_emotions_desc.isEmpty() && s->value("tooltip_show_thoughts",true).toBool())
-        html.append(paragraph.arg(m_emotions_desc));
+        col2.append(paragraph.arg(m_emotions_desc));

     if(!skill_summary.isEmpty())
-        html.append(list_with_header.arg(tr("Skills:")).arg(skill_summary));
+        col1.append(list_with_header.arg(tr("Skills:")).arg(skill_summary));

     if(!m_is_animal && s->value("tooltip_show_mood",false).toBool() && !had_mood()){
         QStringList skill_names;
@@ -2661,20 +2661,20 @@ QString Dwarf::tooltip_text() {
         else {
             skill_names << tr("Craftsdwarf (Bone/Stone/Wood)");
         }
-        html.append(paragraph_with_title.arg(tr("Highest Moodable Skill:")).arg(skill_names.join(",")));
+        col1.append(paragraph_with_title.arg(tr("Highest Moodable Skill:")).arg(skill_names.join(",")));
     }

     if(!personality_summary.isEmpty())
-        html.append(paragraph_with_title.arg(tr("Personality:")).arg(personality_summary));
+        col2.append(paragraph_with_title.arg(tr("Personality:")).arg(personality_summary));

     if(!m_pref_tooltip.isEmpty())
-        html.append(paragraph_with_title.arg(tr("Preferences:")).arg(m_pref_tooltip));
+        col2.append(paragraph_with_title.arg(tr("Preferences:")).arg(m_pref_tooltip));

     if(!roles_summary.isEmpty())
-        html.append(paragraph_with_header.arg(tr("Top %n Roles:", "", max_roles)).arg(roles_summary));
+        col1.append(paragraph_with_header.arg(tr("Top %n Roles:", "", max_roles)).arg(roles_summary));

     if(m_is_animal)
-        html.append(paragraph_with_title.arg(tr("Trained Level:")).arg(get_animal_trained_descriptor(m_animal_type)));
+        col1.append(paragraph_with_title.arg(tr("Trained Level:")).arg(get_animal_trained_descriptor(m_animal_type)));

     if(s->value("tooltip_show_health",false).toBool() && (!m_is_animal || (m_is_animal && s->value("animal_health",false).toBool()))){

@@ -2705,7 +2705,7 @@ QString Dwarf::tooltip_text() {
         }

         if(!health_info.isEmpty())
-            html.append(paragraph.arg(health_info));
+            col1.append(paragraph.arg(health_info));
     }

     if(m_syndromes.count() > 0 && s->value("tooltip_show_buffs",false).toBool()){
@@ -2713,14 +2713,14 @@ QString Dwarf::tooltip_text() {
         QString ailments = get_syndrome_names(false,true);

         if(!buffs.isEmpty())
-            html.append(paragraph_with_title.arg(tr("Buffs:")).arg(buffs));
+            col1.append(paragraph_with_title.arg(tr("Buffs:")).arg(buffs));
         if(!ailments.isEmpty())
-            html.append(paragraph_with_title.arg(tr("Ailments:")).arg(ailments));
+            col1.append(paragraph_with_title.arg(tr("Ailments:")).arg(ailments));
     }


     if(s->value("tooltip_show_caste_desc",true).toBool() && caste_desc() != "")
-        html.append(paragraph.arg(caste_desc()));
+        col1.append(paragraph.arg(caste_desc()));

     if(s->value("highlight_cursed", false).toBool() && m_curse_name != ""){
         QString curse_text = "";
@@ -2735,15 +2735,19 @@ QString Dwarf::tooltip_text() {
                 curse_text.append(tr("born %1 years before the Age of Myth.").arg(abs(m_true_birth_year)));
             }
         }
-        html.append(paragraph.arg(curse_text));
+        col1.append(paragraph.arg(curse_text));
     }

     if(s->value("tooltip_show_kills",false).toBool() && m_hist_figure && m_hist_figure->total_kills() > 0){
-        html.append(paragraph.arg(m_hist_figure->formatted_summary()));
+        col1.append(paragraph.arg(m_hist_figure->formatted_summary()));
     }

     s->endGroup();

+    html.append(QString("<table><tr><td width=\"300\">%1</td><td width=\"600\">%2</td></tr></table>")
+            .arg(col1)
+            .arg(col2));
+
     return html;
 }
```

**Comment by RusTit** [comment]
Ok, then I close the issue.

**Comment by cvuchener** [comment]
Does that you are preparing a PR, or that I should do it?

**Comment by cvuchener** [comment]
OK, close the issue when the patch is merged.

You can start with my patch but you should check the details. I put the details in the columns without much thinking and you should tweak the column width and spacing.

**Comment by cvuchener** [comment]
fixed by #97
